src/ray/object_manager/object_manager.cc

```diff
@@ -164,287 +319,140 @@ ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
   return ray::Status::OK();
 };
 
-ray::Status ObjectManager::GetMsgConnection(
-    const ClientID &client_id, std::function<void(SenderConnection::pointer)> callback) {
-  ray::Status status = Status::OK();
-  if (message_send_connections_.count(client_id) > 0) {
-    callback(message_send_connections_[client_id]);
-  } else {
-    status = object_directory_->GetInformation(
-        client_id,
-        [this, callback](RemoteConnectionInfo info) {
-          Status status = CreateMsgConnection(info, callback);
-        },
-        [this](const Status &status) {
-          // TODO: deal with failure.
-        });
-  }
-  return status;
+void ObjectManager::ProcessNewClient(std::shared_ptr<ReceiverConnection> conn) {
+  conn->ProcessMessages();
 };
 
-ray::Status ObjectManager::CreateMsgConnection(
-    const RemoteConnectionInfo &info,
-    std::function<void(SenderConnection::pointer)> callback) {
-  message_send_connections_.emplace(
-      info.client_id, SenderConnection::Create(io_service_, info.ip, info.port));
-  // Prepare client connection info buffer.
-  flatbuffers::FlatBufferBuilder fbb;
-  bool is_transfer = false;
-  auto message =
-      CreateClientConnectionInfo(fbb, fbb.CreateString(client_id_.binary()), is_transfer);
-  fbb.Finish(message);
-  // Pack into asio buffer.
-  size_t length = fbb.GetSize();
-  std::vector<boost::asio::const_buffer> buffer;
-  buffer.push_back(boost::asio::buffer(&length, sizeof(length)));
-  buffer.push_back(boost::asio::buffer(fbb.GetBufferPointer(), length));
-  // Send synchronously.
-  SenderConnection::pointer conn = message_send_connections_[info.client_id];
-  boost::system::error_code error;
-  boost::asio::write(conn->GetSocket(), buffer);
-  // The connection is ready, invoke callback with connection info.
-  callback(message_send_connections_[info.client_id]);
-  return Status::OK();
+void ObjectManager::ProcessClientMessage(std::shared_ptr<ReceiverConnection> conn,
+                                         int64_t message_type, const uint8_t *message) {
+  switch (message_type) {
+  case OMMessageType_PushRequest: {
+    // RAY_LOG(INFO) << "ProcessClientMessage PushRequest";
+    // TODO(hme): Realize design with transfer requests handled in this manner.
+    break;
+  }
+  case OMMessageType_PullRequest: {
+    // RAY_LOG(INFO) << "ProcessClientMessage PullRequest";
+    ReceivePullRequest(conn, message);
+    conn->ProcessMessages();
+    break;
+  }
+  case OMMessageType_ConnectClient: {
+    // RAY_LOG(INFO) << "ProcessClientMessage ConnectClient";
+    ConnectClient(conn, message);
+    break;
+  }
+  case OMMessageType_DisconnectClient: {
```

Comment:
In many cases, you will not get an actual message when a client disconnects. Let's not assume that you will get any clean information from this case. I would remove the `DisconnectClient` flatbuffer message completely.